### DISCUSSION TITLE
How to (temporary) depend on a dangling commit (from a pull request)?

### DISCUSSION THREAD

**Discussion by ruudk** [composer/composer#11720] [Q&A / Support] [answered]
I want to start testing my CI on PHP 8.3 (not yet released).

I have a dependency on moneyphp/money that is, at the time of writing, not yet compatible with PHP 8.2.

However, there is a PR that adds support: https://github.com/moneyphp/money/pull/757

I know I can depend on this fork, by doing:
```json
{
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/asbiin/money.git",
            "comment": "https://github.com/moneyphp/money/pull/757"
        }
    ],
    "require": {
        "moneyphp/money": "dev-php83"
    }
}
```

But having to create the custom `repositories` is tedious to write.

A much simpler way to write is this:

```json
{
    "require": {
        "moneyphp/money": "dev-master#cea5ff8ef5acc86697c63ef80553679ae5c3a36a"
    }
}
```

This works, because commits associated to a pull request, are also available inside as commits in the target repository.

When you open the commit on GitHub, you would see:
```
This commit does not belong to any branch on this repository, and may belong to a fork outside of the repository.
```
(after I wrote this discussion, the commit was merged, and therefore the message above is gone).

Now when I install this on PHP 8.2, it just works. I verify that I get the forked code.
![Screenshot 2023-11-17 at 13 30 08@2x](https://github.com/composer/composer/assets/104180/2702515e-9076-4169-a310-d67eda6d07ff)

But when I install it on PHP 8.3 it doesn't:
![Screenshot 2023-11-17 at 13 30 21@2x](https://github.com/composer/composer/assets/104180/f7c20804-2c83-46af-a061-7b1907403d9d)

I understand why this is.

I tell composer to install a commit from dev-master. And I think it uses `dev-master` information from packagist.org to do the requirements checking.

Is there a way to tell composer to read the information from the commit? Bypassing dev-master from packagist?

This would make it way easier to (temporary) use a forked commit that's pending merge / tagging.

**Comment by stof** [comment, marked as answer]
The commit-pinning feature is an installer-level hack changing which commit gets downloaded. It has no way to be taken into account by the solver because composer repositories don't create versions for each commit (this would make composer consume TB of memory quite quickly).

If you want to  use a version from a different branch properly, there is no other way than registering the fork as a repository and requiring the version for the branch you want to use.

**Reply by ruudk** [reply]
Thanks, that explains what I'm seeing. 

So basically, we can use the commit-pinning as long as the composer requirements are the same. 

But since this type of PR was changing the composer.json, it was not working.
